### PR TITLE
Additional features and enhanced support for kernel optimisation

### PR DESCRIPTION
**Pull request by FabioLuporini**
This PR:
- Fixes a few bugs related to AST transformation
- Extends the AST hierarchy
- Changes the way padding is performed to handle the case in which offsets are present
- Fixes the expression rewriter when multiple expressions are found in the kernel
- Generally improves the code quality

### CODE REVIEW

**Comment by wence-** [comment]
Does this subsume #36, or does that need to go first?

**Comment by FabioLuporini** [comment]
#36 and the corresponding changes in PyOP2 should go first

File: coffee/utils.py [hunk 1]
```diff
@@ -100,6 +100,70 @@ def unroll_factors(loops):
...
+        def mark(node):
+            if Process.start is not None:
+                Process._processed.append((node, Process.start, Process.end,
+                                           Process.decls, Process.blockable))
+            Process.start = None
+            Process.end = None
+            Process.decls = {}
+            Process.blockable = []
+
+    def init_decl(stmt):
```
Comment (kynan) [inline coffee/utils.py]: Why not make those methods of `Process`?

**Comment by FabioLuporini** [comment]
A tiny note before I forget to say: @wence- , this PR (I've already rebased it) also works around the disgusting hack you had to do for padding in presence of inverse

**Pull request merged**